cpp/include/cudf/detail/nvtx/nvtx.hpp

```diff
@@ -0,0 +1,1810 @@
+   * created through the `Domain::get` function.
+   *
+   * @param name A unique name identifying the domain
+   */
+  explicit Domain(const char* name) noexcept
+      : _domain{nvtxDomainCreateA(name)} {}
+
+  /**
+   * @brief Construct a new Domain with the specified `name`.
+   *
+   * This constructor is private as it is intended that `Domain` objects only be
+   * created through the `Domain::get` function.
+   *
+   * @param name A unique name identifying the domain
+   */
+  explicit Domain(const wchar_t* name) noexcept
+      : _domain{nvtxDomainCreateW(name)} {}
+
+  /**
+   * @brief Construct a new Domain with the specified `name`.
+   *
+   * This constructor is private as it is intended that `Domain` objects only be
+   * created through the `Domain::get` function.
+   *
+   * @param name A unique name identifying the domain
+   */
+  explicit Domain(std::string const& name) noexcept : Domain{name.c_str()} {}
+
+  /**
+   * @brief Construct a new Domain with the specified `name`.
+   *
+   * This constructor is private as it is intended that `Domain` objects only be
+   * created through the `Domain::get` function.
+   *
+   * @param name A unique name identifying the domain
+   */
+  explicit Domain(std::wstring const& name) noexcept : Domain{name.c_str()} {}
+
+  /**
+   * @brief Default constructor creates a `Domain` representing the
+   * "global" NVTX domain.
+   *
+   * All events not associated with a custom `Domain` are grouped in the
+   * "global" NVTX domain.
+   *
+   */
+  Domain() = default;
+
+  /**
+   * @brief Destroy the Domain object, unregistering and freeing all domain
+   * specific resources.
+   */
+  ~Domain() noexcept { nvtxDomainDestroy(_domain); }
+
+ private:
+  nvtxDomainHandle_t const _domain{};  ///< The `Domain`s NVTX handle
+};
+
+/**
+ * @brief Returns reference to the `Domain` object that represents the global
+ * NVTX domain.
+ *
+ * This specialization for `Domain::global` returns a default constructed,
+ * `Domain` object for use when the "global" domain is desired.
+ *
+ * All NVTX events in the global domain across all libraries and applications
+ * will be grouped together.
+ *
+ * @return Reference to the `Domain` corresponding to the global NVTX domain.
+ *
+ */
+template <>
+Domain const& Domain::get<Domain::global>() {
+  static Domain const d{};
+  return d;
+}
+
+/**
+ * @brief Indicates the values of the red, green, blue color channels for
+ * a RGB color code.
+ *
+ */
+struct RGB {
+  /// Type used for component values
+  using component_type = uint8_t;
+
+  /**
+   * @brief Construct a RGB with red, green, and blue channels
+   * specified by `red_`, `green_`, and `blue_`, respectively.
+   *
+   * Valid values are in the range `[0,255]`.
+   *
+   * @param red_ Value of the red channel
+   * @param green_ Value of the green channel
+   * @param blue_ Value of the blue channel
+   */
+  constexpr RGB(component_type red_, component_type green_,
```

Comment:
Normally I use trailing underscores for the member types, but I didn't want to do that in this case because the member vars are intended to be public (didn't want to hassle with adding accessor functions). While I know that the parameter names and member names can be the same, I just think it's clearer to have them be different.